python/ray/serve/_private/result_wrapper.py

```diff
@@ -0,0 +1,148 @@
+import threading
+import time
+from abc import ABC, abstractmethod
+from typing import Callable, Optional, Union
+
+import ray
+from ray.serve._private.utils import calculate_remaining_timeout
+
+
+class ResultWrapper(ABC):
+    @abstractmethod
+    def get(self, timeout_s: Optional[float]):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def get_async(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def next(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def anext(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def add_callback(self, callback: Callable):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def cancel(self):
+        raise NotImplementedError
+
+
+class ActorResultWrapper(ResultWrapper):
+    def __init__(
+        self,
+        obj_ref_or_gen: Union[ray.ObjectRef, ray.ObjectRefGenerator],
+        is_streaming: bool,
+    ):
+        self._obj_ref: Optional[ray.ObjectRef] = None
+        self._obj_ref_gen: Optional[ray.ObjectRefGenerator] = None
+        self._is_streaming: bool = is_streaming
+        self._object_ref_or_gen_sync_lock = threading.Lock()
+
+        if isinstance(obj_ref_or_gen, ray.ObjectRefGenerator):
+            self._obj_ref_gen = obj_ref_or_gen
+        else:
+            self._obj_ref = obj_ref_or_gen
+
+        self._resolved = None
+
+    @property
+    def obj_ref(self) -> Optional[ray.ObjectRef]:
+        return self._obj_ref
+
+    @property
+    def obj_ref_gen(self) -> Optional[ray.ObjectRefGenerator]:
+        return self._obj_ref_gen
+
+    def resolve_gen_to_ref_if_necessary_sync(
+        self, timeout_s: Optional[float] = None
+    ) -> Optional[ray.ObjectRef]:
+        """Returns the object ref pointing to the result."""
+
+        # NOTE(edoakes): this section needs to be guarded with a lock and the resulting
+        # object ref or generator cached in order to avoid calling `__next__()` to
```

Comment:
```suggestion
        # object ref cached in order to avoid calling `__next__()` to
```
I think the comment is outdated?